### PR TITLE
Batch effect correction and deeper models

### PR DESCRIPTION
**Pull request by ben-heil**
This notebook examines the effects of batch effect correcting the data and running deeper models.

Results:
- LIMMA removes all the linear signals that are batch dependent. Because the tissue and study are heavily confounded, this removes all the biological signal models like LR would be able to detect. The nonlinear models can still find something, but performance is much worse
- Deeper models perform less well than the five layer neural network. I'd suspect this is due to vanishing gradients since I haven't added skip-connections or anything. Looks like the five layer network is the sweet spot for this dataset, and I'll avoid anything larger because those models take 5ever to train.

BE corrected binary classification:
image

More models on multiclass:
image

Changes: 
- Wrote new config files for the new models
- Visualized the results in the `benchmark_results` notebook
- Implemented a new class for arbitrarily deep fairly rectangular neural networks in `models.py`
- Modified `predict_tissues` to correct batch effects only on the labeled data because limma/combat wanted ~100s of GBs to batch effect correct the whole dataset

### CODE REVIEW

File: saged/models.py [hunk 1]
```diff
@@ -467,6 +467,58 @@ def set_final_layer(self, new_layer: nn.Module):
...
+
+        self.output = nn.Linear(input_size // 2, output_size)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.l1(x)
+        for layer in self.intermediate:
+            x = layer(x)
+        x = self.output(x)
+
+        return x
```
Comment (jjc2718) [inline saged/models.py:519]: Does the `DeepClassifier` class (below) still have a use? It looks like it's essentially the same as this more general class, but with the number of layers hard-coded instead of passed in as a parameter.

File: notebook/converted/benchmark_results.py [hunk 2]
```diff
@@ -1158,9 +1160,54 @@ def return_unlabeled():
...
+
+tissue_metrics['supervised'] = tissue_metrics['supervised'].str.replace('pytorch_supervised', 'three_layer_net')
+tissue_metrics['supervised'] = tissue_metrics['supervised'].str.replace('deep_net', 'five_layer_net')
+tissue_metrics
+
+
+# In[10]:
+
+
 plot = ggplot(tissue_metrics, aes(x='train_count', y='balanced_accuracy', color='supervised'))
```
Comment (jjc2718) [inline notebook/converted/benchmark_results.py:1211]: It's definitely interesting that the three-layer and seven-layer networks perform so much worse than the five-layer net. Maybe the learning rate or other hyperparameters could be affecting this? I guess it makes sense to avoid deeper networks if you're fairly sure it won't be an easy fix.

**Review by jjc2718 — APPROVED** [review]
Looks good to me!

File: saged/models.py (hunk 1 above)
Comment (ben-heil) [reply saged/models.py:519]: Yeah, the five layer network uses the DeepClassifier class as a base

File: notebook/converted/benchmark_results.py (hunk 2 above)
Comment (ben-heil) [reply notebook/converted/benchmark_results.py:1211]: Yeah for sure. I think it's some combination of being a good depth for the problem space and not running into vanishing gradients.

File: saged/models.py (hunk 1 above)
Comment (jjc2718) [reply saged/models.py:519]: Got it! Might be good to have them use the same base class in the future (that way if you make changes to one you won't have to remember to also make changes to the other), but I'm fine with it for now.
